Modify Multi-network annotation generation.

mrthinger opened the pull request:
NOTE: This PR is originally from #532, I will be owning its review and making necessary modifications to get it merged based on comments. I have reviewed the original PR and have approved.

This change accounts for Network Ready condition and Node's network-status annotation.
The north-interface annotation is set only for Ready Networks. The networks annotation and IP capacity is set for networks in network-status annotation.

Added new reconcile triggers:
- Node MN network-status annotation change
- Network Ready condition state change

sdmodi commented:
Evan would you please resolve the merge conflict and put this up for review. I'll take an initial look at it today. Thanks!

sdmodi commented on pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go:
```diff
@@ -26,13 +30,28 @@ func (ca *cloudCIDRAllocator) performMultiNetworkCIDRAllocation(node *v1.Node, i
 			networks = append(networks, network)
+			_, ok := nodeUpNet[network.Name]
+			if ok {
+				// we do not care about status of default Network, since it has to be
+				// always available
```
The comment seems to be at the wrong place. If I read this code correctly, the default network will never get here since nodeUpNet will never have the default network. We should move this comment out of the if statement.

mrthinger replied:
I've modified the comment to be more clear. Let me know if that works.

sdmodi commented on pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go:
```diff
@@ -42,6 +61,11 @@ func (ca *cloudCIDRAllocator) performMultiNetworkCIDRAllocation(node *v1.Node, i
 			rangeNameAliasIPMap[ipRange.SubnetworkRangeName] = ipRange
 		}
 		for _, network := range networks {
+			if _, ok := matchedNetworks[network.Name]; ok {
+				// skip networks that are already matched with an interface
+				continue
+			}
+
```
I didn't get when we would hit this condition and continue.

shouri007 replied:
since outer loop is at interface level, (inf0 -- {nw1,nw2,nw3}) and let's say inf0 matched with nw3, we don't need to check for nw3 when we process next interface (inf1)

sdmodi commented on pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator_test.go:
```diff
@@ -15,7 +15,12 @@ const (
 )
 
-func network(name, gkeNetworkParamsName string) *networkv1.Network {
+func network(name, gkeNetworkParamsName string, isReady bool) *networkv1.Network {
```
Did you intend to add a unit test for getUpNetworks?

mrthinger replied:
I'm not sure what Maciej originally intended, but I have added a unit test for getUpNetworks.

shouri007 commented on pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go:
```diff
@@ -26,13 +30,28 @@ func (ca *cloudCIDRAllocator) performMultiNetworkCIDRAllocation(node *v1.Node, i
+	nodeUpNet, err := getUpNetworks(node)
+
+	// statusNetworks is map of networks that are Ready and status==UP
+	statusNetworks := make(map[string]struct{})
```
call this `readyNetworks`?

mrthinger replied:
I renamed readyAndUpNetworks. Its a bit more verbose, but readyNetworks would've specifically referred to the network object's ready condition.

shouri007 commented on pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go:
```diff
@@ -26,13 +30,28 @@ func (ca *cloudCIDRAllocator) performMultiNetworkCIDRAllocation(node *v1.Node, i
 		}
 	}
+
+	matchedNetworks := make(map[string]struct{})
```
call this `processedNws`?

shouri007 commented on pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go:
```diff
@@ -77,7 +102,9 @@ func (ca *cloudCIDRAllocator) performMultiNetworkCIDRAllocation(node *v1.Node, i
 					}
 				} else {
 					northInterfaces = append(northInterfaces, networkv1.NorthInterface{Network: network.Name, IpAddress: inf.NetworkIP})
```
should we push northInterfaces also within the `statusNetworks` check below? @sdmodi wdyt?

sdmodi replied:
No we should not. Without northInterface populated, anetd will never set network status to ready. The question really was, whether we should do both northInterface and networks annotation at the same time and we decided not to do so.

MrHohn commented:
/ok-to-test

sdmodi commented:
@mrthinger would you please reach out to Antonio for a review and approval once you have addressed mine and Shouri's comments. Thanks!

aojea commented on pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go:
```diff
@@ -26,13 +30,28 @@ func (ca *cloudCIDRAllocator) performMultiNetworkCIDRAllocation(node *v1.Node, i
 	}
+
+	// get networks from Node's network-status annotation
+	nodeUpNet, err := getUpNetworks(node)
```
don't you check the error?

mrthinger replied:
added.

aojea commented on pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go:
```diff
@@ -26,13 +30,28 @@ func (ca *cloudCIDRAllocator) performMultiNetworkCIDRAllocation(node *v1.Node, i
-	// ignore networks that are under deletion.
+	// filter networks based on Ready condition
 	for _, network := range k8sNetworksList {
-		if network.DeletionTimestamp.IsZero() {
```
you may want to keep skipping the networks that are being deleted
```
if !network.DeletionTimestamp.IsZero() {
   continue
```

mrthinger replied:
We are shifting to using the network-status annotation as the source of truth for whether a network is up or down as opposed to watching the deletion timestamp. In this case, we could technically use the deletion timestamp to bypass the source of truth at the current moment, but the dependencies for a network being up or down status may change in the future and this keeps things simple as all components rely on status.

sdmodi replied:
Evan, is it the case that if the network is being deleted the NetworkConditionStatusReady will be false? Otherwise we could do the optimization Antonio is suggesting.

mrthinger replied:
NetworkConditionStatusReady wont be false if the network is being deleted. We could do the optimization now. If something changes in the future where we need to keep the annotation until network status changes we could change the behavior here.

mrthinger replied:
added

aojea commented:
```
pkg/controller/nodeipam/ipam/cloud_cidr_allocator_test.go:1274:29: not enough arguments in call to network
	have (string, string)
	want (string, string, bool)
```

mrthinger commented:
> ```
> pkg/controller/nodeipam/ipam/cloud_cidr_allocator_test.go:1274:29: not enough arguments in call to network
> 	have (string, string)
> 	want (string, string, bool)
> ```

https://github.com/kubernetes/cloud-provider-gcp/pull/534 got merged into master after this PR which duplicated roughly 500 lines of tests. So the automated jobs were testing the merged version of this PR with those changes included.

I rebased this PR and fixed the issue. I was able to combine the metrics tests as an optional assertion to the original set of tests and getting rid of duplicated code.

mrthinger commented on pkg/controller/nodeipam/ipam/cloud_cidr_allocator_test.go:
```diff
@@ -970,240 +1004,39 @@ func TestUpdateCIDRAllocation(t *testing.T) {
 			},
 			nodeChanges: func(node *v1.Node) {},
 		},
-	}
```
the diff goes absolutely nuts here. Its probably better to view this file in an editor.

This is where I was able to delete the duplicated test from #534

sdmodi commented on pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go:
```diff
@@ -26,13 +30,31 @@ func (ca *cloudCIDRAllocator) performMultiNetworkCIDRAllocation(node *v1.Node, i
+	}
+
+	// readyAndUpNetworks is map of networks that are Ready and status==UP
+	readyAndUpNetworks := make(map[string]struct{})
```
Thinking about this a bit more, under what condition would a network be in the 'upStatusnetworks' list but its network.Status.Conditions is not ready? In other words, if the network was annotated on the node, the network object must have been in a ready state.

mrthinger replied:
I think youre right. If status is up, the network has to be ready. I don't think there is an issue having both checks since it doesnt cost us anything extra to compute.

Do you think we should remove the Network Ready condition from this and only have it be status==up?

sdmodi replied:
Yes. I think we should remove it and make the code simpler. I am less worried about the CPU utilization and more worried about code readability. This is much simpler to reason about if we only worry about the network status annotation.

sdmodi reviewed (commented):
Antonio, would you please review this today. We should check this in and create a CCM build for GKE usage ASAP. Thanks!

aojea commented:
/lgtm
/approve
